lib/game/game_engine.rb

```diff
@@ -8,11 +10,35 @@ def initialize(ui, board_size = 3)
     def start(player_one, player_two)
       game = Game.new([player_one, player_two], board_size: @board_size)
       @ui.show_game_state(game)
+      game = play_game(game)
+      @ui.show_game_result(game)
+    end
+
+    private
+
+    def play_game(game)
       while !game.game_over?
-        game = game.make_move(@ui)
+        game = play_a_turn(game)
         @ui.show_game_state(game)
       end
-      @ui.show_game_result(game)
+      game
+    end
+
+    def play_a_turn(game)
+      loop do
+        made_successful_move, game = make_move(game)
+        break if made_successful_move
+      end
+      game
+    end
+
+    def make_move(game)
+      begin
+        [true, game.make_move(@ui)]
+      rescue InvalidPositionError
```

Comment:
https://stackoverflow.com/questions/1542672/how-does-one-use-rescue-in-ruby-without-the-begin-and-end-block .... to get rid of the extra blocks, you can just have the rescue since you're having this go for the entire function.

Reply:
Also helpful, thanks. Plus I realised that I don't need to return true or false, I can just return the new `game` object returned by `make_move` and break out of the loop in the caller if `new_game != old_game`.